documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1189 @@
+
+    public static Vector2F<TScalar> operator *(TScalar left, Vector2F<TScalar> right);
+
+    public static Vector2F<TScalar> operator +(Vector2F<TScalar> value);
+    public static Vector2F<TScalar> operator -(Vector2F<TScalar> value);
+
+    public TScalar LengthSquared { get; }
+    public TScalar Length();
+
+    public Vector2F<TScalar> Dot(Vector2F<TScalar> other);
+    public TScalar DistanceSquared(Vector2F<TScalar> other);
+    public TScalar Distance(Vector2F<TScalar> other);
+    public Vector2F<TScalar> Abs();
+    public void CopyTo(TScalar[] array);
+    public void CopyTo(TScalar[] array, int index);
+
+    public static Vector2F<TScalar> Dot (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> DistanceSquared (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> Distance (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> Clamp (Vector2F<TScalar> value, Vector2F<TScalar> min, Vector2F<TScalar> max);
+    public static Vector2F<TScalar> Min (Vector2F<TScalar> value1, Vector2F<TScalar> value2);
+    public static Vector2F<TScalar> Max (Vector2F<TScalar> value1, Vector2F<TScalar> value2);
+    public static Vector2F<TScalar> Abs (Vector2F<TScalar> value);
+}
+```
+At the end of this proposal is the definition for the new interfaces used in the these classes.
+
+# Matrix Types
+
+This proposal includes the following matrix types:
+- Matrix2x2F
+- Matrix2x2I
+- Matrix2x3F
+- Matrix2x3I
+- Matrix2x4F
+- Matrix2x4I
+- Matrix3x2F
+- Matrix3x2I
+- Matrix3x3F
+- Matrix3x3I
+- Matrix3x4F
+- Matrix3x4I
+- Matrix4x2F
+- Matrix4x2I
+- Matrix4x3F
+- Matrix4x3I
+- Matrix4x4F
+- Matrix4x4I
+- Matrix5x4F
+- Matrix5x4I
+
+Integer Variants do not require any functions which interact with Quaternions
+
+Matricies **must** fulfill the following requirements:
+- be a struct
+- Stored in row major format
+- F matricies work with F vectors, and I Matricies work with I vectors
+- Both column and row vectors and individual values accessible via properties
+- An indexer that takes row and column indicies and outputs the value
+- Add, subtract, and multiply operators defined with Matricies of the same size
+- Multiply operators defined with compatible matricies, if the output matrix type already exists (AxB * BxC = AxC)
+- Multiply operator defined with the appropriate Vector type (vector4F * Matrix4x4F is valid)
+  - **optionally**, include multiplication operators for Vector types smaller than the appropriate type, assuming 1 for the missing dimensions (vector2F * Matrix4x4F -> (Vector2F.x, Vector2F.y, 1, 1) * Matrix4x4F)
+- Negate Operator defined
+- Fulfills `IEquatable<T>` where `T` is the same matrix class and the System.Numerics version of the matrix if available
+- Implicit conversion to and from the System.Numerics matrix type, if available
+- Invert function for square matricies
+- GetDeterminant function for square matricies and Matrix3x2
```

Comment:
Shouldn't this also be for `Matrix4x3` and `Matrix5x4`? Much like `Matrix3x2`, they exist as a memory/perf optimization since the last column can be inferred to be `0, 0, 0, 1` (since matrices expanded up take on the same positional value as `Identity` would have)

Reply:
👍